fem/hyperbolic_conservation_laws.cpp

```diff
@@ -0,0 +1,493 @@
+   const double density = U(0);                  // ρ
+   const Vector momentum(U.GetData() + 1, dim);  // ρu
+   const double energy = U(1 + dim);             // E, internal energy ρe
+   // pressure, p = (γ-1)*(ρu - ½ρ|u|²)
+   const double pressure = (specific_heat_ratio - 1.0) *
+                           (energy - 0.5 * (momentum * momentum) / density);
+
+   // Check whether the solution is physical only in debug mode
+   MFEM_ASSERT(density >= 0, "Negative Density");
+   MFEM_ASSERT(pressure >= 0, "Negative Pressure");
+   MFEM_ASSERT(energy >= 0, "Negative Energy");
+
+   // 2. Compute Flux
+   for (int d = 0; d < dim; d++)
+   {
+      FU(0, d) = momentum(d);  // ρu
+      for (int i = 0; i < dim; i++)
+      {
+         // ρuuᵀ
+         FU(1 + i, d) = momentum(i) * momentum(d) / density;
+      }
+      // (ρuuᵀ) + p
+      FU(1 + d, d) += pressure;
+   }
+   // enthalpy H = e + p/ρ = (E + p)/ρ
+   const double H = (energy + pressure) / density;
+   for (int d = 0; d < dim; d++)
+   {
+      // u(E+p) = ρu*(E + p)/ρ = ρu*H
+      FU(1 + dim, d) = momentum(d) * H;
+   }
+
+   // 3. Compute maximum characteristic speed
+
+   // sound speed, √(γ p / ρ)
+   const double sound = sqrt(specific_heat_ratio * pressure / density);
+   // fluid speed |u|
+   const double speed = sqrt(momentum * momentum) / density;
+   // max characteristic speed = fluid speed + sound speed
+   return speed + sound;
+}
+
+
+double EulerFlux::ComputeFluxDotN(const Vector &x,
+                                  const Vector &normal,
+                                  FaceElementTransformations &Tr, Vector &FUdotN) const
+{
+   const int dim = normal.Size();
+
+   // 1. Get states
+   const double density = x(0);                  // ρ
+   const Vector momentum(x.GetData() + 1, dim);  // ρu
+   const double energy = x(1 + dim);             // E, internal energy ρe
+   // pressure, p = (γ-1)*(E - ½ρ|u|^2)
+   const double pressure = (specific_heat_ratio - 1.0) *
+                           (energy - 0.5 * (momentum * momentum) / density);
+
+   // Check whether the solution is physical only in debug mode
+   MFEM_ASSERT(density >= 0, "Negative Density");
+   MFEM_ASSERT(pressure >= 0, "Negative Pressure");
+   MFEM_ASSERT(energy >= 0, "Negative Energy");
+
+   // 2. Compute normal flux
+
+   FUdotN(0) = momentum * normal;  // ρu⋅n
+   // u⋅n
+   const double normal_velocity = FUdotN(0) / density;
+   for (int d = 0; d < dim; d++)
+   {
+      // (ρuuᵀ + pI)n = ρu*(u⋅n) + pn
+      FUdotN(1 + d) = normal_velocity * momentum(d) + pressure * normal(d);
+   }
+   // (u⋅n)(E + p)
+   FUdotN(1 + dim) = normal_velocity * (energy + pressure);
+
+   // 3. Compute maximum characteristic speed
+
+   // sound speed, √(γ p / ρ)
+   const double sound = sqrt(specific_heat_ratio * pressure / density);
+   // fluid speed |u|
+   const double speed = sqrt(momentum * momentum) / density;
+   // max characteristic speed = fluid speed + sound speed
+   return speed + sound;
```

Comment:
This wavespeed estimate uses the vector velocity magnitude instead of the normal velocity magnitude which gets used by the Riemann solver. It's not necessarily wrong but the second one is more typical (Davis wavespeed estimate) for Rusanov-type Riemann solvers.